js_modules/dagster-ui/packages/ui-core/src/pipelines/JobTabs.tsx

```diff
@@ -0,0 +1,117 @@
+  isJob: boolean;
+  explorerPath: ExplorerPath;
+  matchingTab?: string;
+  permissions: PermissionsState;
+  tabs: JobTabConfig[];
+}
+
+export const JobTabs = (props: Props) => {
+  const {repoAddress, isJob, explorerPath, matchingTab = '', permissions, tabs} = props;
+
+  const explorerPathForTab = explorerPathToString({
+    ...explorerPath,
+    opNames: [],
+  });
+
+  const selectedTab = React.useMemo(() => {
+    return (
+      tabs.find((tab) => tab.pathComponent === matchingTab) ||
+      tabs.find((tab) => tab.pathComponent === '')
+    );
+  }, [matchingTab, tabs]);
+
+  return (
+    <Tabs size="large" selectedTabId={selectedTab!.id}>
+      {tabs
+        .filter((tab) => !tab.isHidden)
+        .map((tab) => {
+          const {id, title: text, getPermissionsResult} = tab;
+          const permissionsResult = getPermissionsResult ? getPermissionsResult(permissions) : null;
+          const disabled = !!(permissionsResult && !permissionsResult.enabled);
+          const title =
+            permissionsResult && disabled ? (
+              <Tooltip content={permissionsResult.disabledReason} placement="top">
+                {text}
+              </Tooltip>
+            ) : (
+              text
+            );
+
+          const href = workspacePathFromAddress(
+            repoAddress,
+            `/${isJob ? 'jobs' : 'pipelines'}/${explorerPathForTab}${tab.pathComponent}`,
+          );
+
+          if (disabled) {
+            return <Tab disabled key={id} id={id} title={title} />;
+          }
+
+          return <TabLink key={id} id={id} title={title} disabled={disabled} to={href} />;
+        })}
+    </Tabs>
+  );
+};
+
+export type JobTabConfigInput = {
+  hasLaunchpad: boolean;
+  hasPartitionSet: boolean;
+};
+
+export interface JobTabConfig {
+  id: string;
+  title: string;
+  pathComponent: string;
+  getPermissionsResult?: (permissionsState: PermissionsState) => PermissionResult;
+  isHidden?: boolean;
+}
+
+export const buildJobTabMap = (input: JobTabConfigInput): Record<string, JobTabConfig> => {
+  const {hasLaunchpad, hasPartitionSet} = input;
+  return {
+    overview: {
+      id: 'overview',
+      title: 'Overview',
+      pathComponent: '',
+    },
+    playground: {
+      id: 'launchpad',
+      title: 'Launchpad',
+      pathComponent: 'playground',
+      getPermissionsResult: (permissionsState: PermissionsState) =>
+        permissionResultForKey(permissionsState, 'canLaunchPipelineExecution'),
+      isHidden: !hasLaunchpad,
+    },
+    runs: {
+      id: 'runs',
+      title: 'Runs',
+      pathComponent: 'runs',
+    },
+    partitions: {
+      id: 'partitions',
+      title: 'Partitions',
+      pathComponent: 'partitions',
+      isHidden: !hasPartitionSet,
+    },
+  };
+};
+
+export const buildJobTabs = (input: JobTabConfigInput): JobTabConfig[] => {
+  const tabConfigs = buildJobTabMap(input);
+  return DEFAULT_JOB_TAB_ORDER.map((tabId) => tabConfigs[tabId]).filter(
```

Comment:
It may make sense based on the usage of `buildJobTabMap` in cloud, but offhand it seems like having `buildJobTabMap` return an ordered array would be better than a map + an ordering array?

Reply:
The idea here was to be able to reorder the tabs if necessary in Cloud. I felt like it would be more annoying to have to reorder/insert on the array rather than just provide a map of values and allow the overriding context to define a fresh order.